### PR TITLE
Do not omit parentheses for numeric literal with ".property" if it can not be followed by dot operator directly.

### PR DESCRIPTION
**Pull request by vilic**
Fix issue #4603.

Added a function `getEmittingNumericLiteralText` as a helper of determine whether we want to keep the parentheses, as the emitted numeric literal maybe different under different target.

E.g., `0o123` emits `83` (which needs the parentheses) under ES3/5, but emits `0o123` under ES6.

### CODE REVIEW

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -897,6 +897,16 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
                     write(text);
                 }
             }
+            
+            function getEmittingNumericLiteralText(node: LiteralExpression): string {
+                let text = getLiteralText(node);
+                
+                if (languageVersion < ScriptTarget.ES6 && isBinaryOrOctalIntegerLiteral(node, text)) {
+                    return node.text;
+                } else {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: `else` on the next line

**Comment by DanielRosenwasser** [comment]
Personally, I'd rather we just always emit the parens. Thoughts @mhegazy?

**Comment by mhegazy** [comment]
We try to emit clean code that you would write by hand. and i do not think you will always parenthesize your numeric literals.

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -2352,7 +2362,8 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
                             operand.kind !== SyntaxKind.PostfixUnaryExpression &&
                             operand.kind !== SyntaxKind.NewExpression &&
                             !(operand.kind === SyntaxKind.CallExpression && node.parent.kind === SyntaxKind.NewExpression) &&
-                            !(operand.kind === SyntaxKind.FunctionExpression && node.parent.kind === SyntaxKind.CallExpression)) {
+                            !(operand.kind === SyntaxKind.FunctionExpression && node.parent.kind === SyntaxKind.CallExpression) &&
+                            !(operand.kind === SyntaxKind.NumericLiteral && node.parent.kind === SyntaxKind.PropertyAccessExpression && !/^0[box]|[e.]/i.test(getEmittingNumericLiteralText(<LiteralExpression>operand)))) {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: This doesn't cover when your numeric literal consists of escapes sequences.

**Comment by DanielRosenwasser** [comment]
@mhegazy I'd rather we just didn't special case this. `2.0.toString()` doesn't seem more canonical to me than `(2).toString()`

File: src/compiler/emitter.ts (hunk 2 above)
Comment (vilic) [inline src/compiler/emitter.ts]: I didn't notice there could be escape sequences in a numeric literal. ES6 stuff?

Comment (vilic) [inline src/compiler/emitter.ts]: So what do you mean by "numeric literal consists of escape sequences"?

Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: You can literally write `var x = \u0032\u002e\u0030` instead of `var x = 2.0` and both are valid and ECMAScript code.

Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Actually, I stand corrected - the escapes can't be used in numeric literals. From [10.1 - Source Text](http://www.ecma-international.org/ecma-262/6.0/#sec-source-text):

> In string literals, regular expression literals, template literals and identifiers, any Unicode code point may also be expressed using Unicode escape sequences that explicitly express a code point’s numeric value. Within a comment, such an escape sequence is effectively ignored as part of the comment.

But as you mentioned, it doesn't

Comment (vilic) [inline src/compiler/emitter.ts]: @DanielRosenwasser :+1:

**Comment by mhegazy** [comment]
so @DanielRosenwasser what is the proposal, just check if the expression is numeric literal, and the parent is property access expression? this sound fine by me. @vilic, any concerns?

**Comment by DanielRosenwasser** [comment]
@mhegazy that sounds fine, or maybe even just don't strip parens from a numeric literal in a type assertion because it's likely going to have a property access for whatever reason anyway. Either is fine with me.

**Comment by vilic** [comment]
@mhegazy @DanielRosenwasser I am cool with both. But I prefer keeping property access as a condition though other situations might be rare.

**Comment by DanielRosenwasser** [comment]
:+1:

**Comment by DanielRosenwasser** [comment]
Thanks @vilic!

**Pull request merged**